src/Http/Http.Abstractions/src/Internal/PathStringHelper.cs

```diff
@@ -1,47 +1,64 @@
-            false, true,  true,  true,  true,  true,  true,  true,      // 0x60 - 0x67
-            true,  true,  true,  true,  true,  true,  true,  true,      // 0x68 - 0x6F
-            true,  true,  true,  true,  true,  true,  true,  true,      // 0x70 - 0x77
-            true,  true,  true,  false, false, false, true,  false,     // 0x78 - 0x7F
+        private static readonly uint[] ValidPathChars = {
+            0b_0000_0000__0000_0000__0000_0000__0000_0000, // 0x00 - 0x1F
+            0b_0010_1111__1111_1111__1111_1111__1101_0010, // 0x20 - 0x3F
+            0b_1000_0111__1111_1111__1111_1111__1111_1111, // 0x40 - 0x5F
+            0b_0100_0111__1111_1111__1111_1111__1111_1110, // 0x60 - 0x7F
         };
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool IsValidPathChar(char c)
         {
-            return c < ValidPathChars.Length && ValidPathChars[c];
+            // Use local array and uint .Length compare to elide the bounds check on array access
+            var validChars = ValidPathChars;
+            var i = (int)c;
+
+            // Array is in chunks of 32 bits, so get offset by dividing by 32
+            var offset = i / 32;
+            // Significant bit position is the remainder of the above calc; i % 32 => i & 31
+            var significantBit = 1u << (i & 31);
+
+            // Check offset in bounds and check if significant bit set
+            return (uint)offset < (uint)validChars.Length &&
+                ((validChars[offset] & significantBit) == significantBit);
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool IsPercentEncodedChar(string str, int index)
         {
-            return index < str.Length - 2
-                && str[index] == '%'
-                && IsHexadecimalChar(str[index + 1])
-                && IsHexadecimalChar(str[index + 2]);
+            var len = (uint)str.Length;
+            if (str[index] == '%' && index < len - 2)
+            {
+                return AreFollowingTwoCharsHex(str, index);
+            }
+
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static bool AreFollowingTwoCharsHex(string str, int index)
+        {
+            Debug.Assert(index < str.Length - 2);
+
+            var c1 = str[index + 1];
+            var c2 = str[index + 2];
+            return IsHexadecimalChar(c1) && IsHexadecimalChar(c2);
         }
 
-        public static bool IsHexadecimalChar(char c)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsHexadecimalChar(char c)
         {
-            return ('0' <= c && c <= '9')
-                || ('A' <= c && c <= 'F')
-                || ('a' <= c && c <= 'f');
+            // Between 0 - 9 or uppercased between A - F
+            return (c - '0') <= 9 || ((c & ~0x20) - 'A') <= ('F' - 'A');
```

Comment:
The casts to an unsigned is needed to correctly handle values <= '0' (now this will result in `true`, e.g. for `\0 - '0'` the first condition is met). Analogous for the seconds condition.

In the proposal I had `byte`, but this isn't enough to cover the whole char-range.
```suggestion
            return (uint)(c - '0') <= 9 || (uint)((c & ~0x20) - 'A') <= ('F' - 'A');
```

So the following test will pass:
```c#
[Fact]
public void IsHexadecimalChar_all_chars()
{
    for (char c = char.MinValue; c < (char)char.MaxValue; ++c)
    {
        bool expected = ('0' <= c && c <= '9') || ('A' <= c && c <= 'F') || ('a' <= c && c <= 'f');
        bool actual = IsHexadecimalChar(c);

        Assert.Equal(expected, actual);
    }
}
```